Detectors/CTP/simulation/src/Digits2Raw.cxx

```diff
@@ -0,0 +1,181 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file Digits2Raw.cxx
+/// \author Roman Lietava
+
+#include "CTPSimulation/Digits2Raw.h"
+#include "FairLogger.h"
+#include "CommonUtils/StringUtils.h"
+
+using namespace o2::ctp;
+
+void Digits2Raw::init()
+{
+  // CTP Configuration
+  if (mCCDBServer.empty()) {
+    LOG(FATAL) << "CTP digitizer: CCDB server is not set";
+  } else {
+    LOG(INFO) << "CTP digitizer:: CCDB server:" << mCCDBServer;
+  }
+  auto& mgr = o2::ccdb::BasicCCDBManager::instance();
+  mgr.setURL(mCCDBServer);
+  mCTPConfiguration = mgr.get<CTPConfiguration>(o2::ctp::CCDBPathCTPConfig);
+  //
+  // Register links
+  //
+  std::string outd = mOutDir;
+  if (outd.back() != '/') {
+    outd += '/';
+  }
+  for (int ilink = 0; ilink < mNLinks; ilink++) {
+    mFeeID = uint64_t(ilink);
+    std::string outFileLink = mOutputPerLink ? o2::utils::Str::concat_string(outd, "ctp_link", std::to_string(ilink), ".raw") : o2::utils::Str::concat_string(outd, "ctp.raw");
+    mWriter.registerLink(mFeeID, mCruID, ilink, mEndPointID, outFileLink);
+  }
+  mWriter.setEmptyPageCallBack(this);
+}
+void Digits2Raw::processDigits(const std::string& fileDigitsName)
+{
+  std::unique_ptr<TFile> digiFile(TFile::Open(fileDigitsName.c_str()));
+  if (!digiFile || digiFile->IsZombie()) {
+    LOG(FATAL) << "Failed to open input digits file " << fileDigitsName;
+    return;
+  }
+  TTree* digiTree = (TTree*)digiFile->Get("o2sim");
+  if (!digiTree) {
+    LOG(FATAL) << "Failed to get digits tree";
+    return;
+  }
+  std::vector<o2::ctp::CTPDigit> CTPDigits, *fCTPDigitsPtr = &CTPDigits;
+  if (digiTree->GetBranch("CTPDigits")) {
+    digiTree->SetBranchAddress("CTPDigits", &fCTPDigitsPtr);
+  } else {
+    LOG(FATAL) << "Branch CTPDigits is missing";
+    return;
+  }
+  // Get first orbit
+  digiTree->GetEntry(0);
+  uint32_t orbit0;
+  if (CTPDigits.size() > 0) {
+    orbit0 = CTPDigits[0].intRecord.orbit;
+  } else {
+    LOG(FATAL) << "Branch CTPDigits there but no entries";
+    return;
+  }
+  // Add all CTPdigits for given orbit
+  for (int ient = 1; ient < digiTree->GetEntries(); ient++) {
```

Comment:
therefore you should start from 0 (upper bound `digiTree->GetEntries()` does not hurt, but you have just 1 entry).

Reply:
@shahor02 So can I assume that  I expect one entry, i.e. more than 1 entry is error ?

Reply:
Currently it is 1 entry, but better do not depend on this assumption, can't you loop over all entries?

Reply:
yes, ok, just asking. I'll loop.